### PR TITLE
Theming options for Gtk and Qt

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

In place of #70276

When I started to use NixOS at desktop I was wondered that there is no module to set Gtk and Qt widget and icon themes (in bare WM environment). Its no joy to use lxappearence, qt5ct and similar tools. These modules fix that issue by adding a way to configure themes declaratively.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @worldofpeace 

TODO: ping jtojnar mid february

### CODE REVIEW

File: nixos/modules/config/qt.nix [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+      assertions = [
+        {
+          assertion = cfg.font == null && cfg.iconTheme == null;
+          message = "`qt.font` and `qt.iconTheme` are only supported by kde platform.";
+        }
+      ];
+      environment.variables.QT_QPA_PLATFORMTHEME = "qgnomeplatform";
+      # TODO: make this optional once https://github.com/NixOS/nixpkgs/issues/54150 is fixed
+      # qgnomeplatform reads theme and other settings from dconf db
+      environment.variables.QT_STYLE_OVERRIDE = cfg.theme.name;
```
Comment (gnidorah) [inline nixos/modules/config/qt.nix]: @worldofpeace In the past it was limited to `adwaita`, but actually `qgnomeplatform` allows any theme to use

File: nixos/modules/config/gtk/gtk.nix [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+  toGtk2Ini = key: value:
+    let
+      value' =
+        if isBool value then (if value then "true" else "false")
+        else if isString value then "\"${value}\""
+        else toString value;
+    in
+      "${key} = ${value'}";
+  toGtk3Ini = generators.toINI {
```
Comment (jtojnar) [inline nixos/modules/config/gtk/gtk.nix]: GTK3 uses GKeyFile, not ini: https://gitlab.gnome.org/GNOME/gtk/blob/3.24.12/gtk/gtksettings.c#L75-86

File: nixos/modules/config/gtk/gtk.nix [hunk 3]
```diff
@@ -0,0 +1,146 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.gtk;
+  gtk2 = cfg.enable && cfg.gtk2;
+
+  toGtk2Ini = key: value:
```
Comment (jtojnar) [inline nixos/modules/config/gtk/gtk.nix]: This is not an ini file either: https://gitlab.gnome.org/GNOME/gtk/blob/2.24.31/docs/reference/gtk/tmpl/gtkrc.sgml

File: nixos/modules/config/qt.nix (hunk 1 above)
Comment (worldofpeace) [reply nixos/modules/config/qt.nix]: Are we sure, when I tested it by infinisil's request I got segfaults...

Comment (gnidorah) [reply nixos/modules/config/qt.nix]: @worldofpeace It works for me with other themes like breeze

File: nixos/modules/rename.nix [hunk 4]
```diff
@@ -169,6 +169,14 @@ with lib;
     # KDE Plasma 5
     (mkRenamedOptionModule [ "services" "xserver" "desktopManager" "kde5" ] [ "services" "xserver" "desktopManager" "plasma5" ])
 
+    # Qt
+    (mkRemovedOptionModule [ "qt5" "style" ] "Please use <option>qt5.theme</option> instead.")
```
Comment (worldofpeace) [inline nixos/modules/rename.nix]: but it seems qt ubiquitously refers to them as styles?

File: nixos/modules/config/gtk/gtk.nix (hunk 2 above)
Comment (gnidorah) [reply nixos/modules/config/gtk/gtk.nix]: @jtojnar Renamed. If you're about implementation, it was borrowed from @rycee https://github.com/rycee/home-manager/blob/master/modules/misc/gtk.nix `home-manager` gtk module and works fine for people

File: nixos/modules/rename.nix (hunk 4 above)
Comment (gnidorah) [reply nixos/modules/rename.nix]: @worldofpeace Fixed, thanks!

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Could we configure cursor themes for gtk as well?
I believe it's just another key/s in gtkrc.

**Comment by gnidorah** [comment]
@worldofpeace 

> Could we configure cursor themes for gtk as well?
I believe it's just another key/s in gtkrc.

Done, thanks!

File: nixos/modules/config/gtk/gtk.nix [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        '';
+        description = "The GTK+ theme to use.";
+      };
+    };
+  };
+
+  config = mkMerge [
+
+    (mkIf gtk2 {
+      environment.systemPackages = [ pkgs.gtk2 ];
```
Comment (worldofpeace) [inline nixos/modules/config/gtk/gtk.nix]: Why do we need gtk2 in systemPackages?

Comment (gnidorah) [reply nixos/modules/config/gtk/gtk.nix]: @worldofpeace Without this, there was an error "Unable to locate theme engine in module_path: "pixmap"" in the past when using e.g. lxappearance. I can't reproduce it, so will remove it for now.

**Comment by gnidorah** [comment]
@worldofpeace @jtojnar merge please?

**Comment by gnidorah** [comment]
@wordofpeace @jtojnar ping

**Comment by rnhmjoj** [comment]
@gnidorah Could you expand the manual chapter "GTK/Qt themes" with a mention to this module and an example?

**Comment by gnidorah** [comment]
@rnhmjoj Done, thank you!

**Review by rnhmjoj — APPROVED** [review]
I've tested every option of `gtk` and I can completely reproduce my current configuration with it.
Also the example works as expected. I tried another Qt style (arc-kde) but I can't get to work, it's probably my fault: I know nothing of Qt.

IMHO this has waited long enough, I hoped it would be in 20.03 already. Let's merge it.

**Comment by gnidorah** [comment]
@rnhmjoj Thank you! arc-kde doesn't provide native qt style, instead it provides settings for 3rd party theme engine (kvantum)

**Comment by rnhmjoj** [comment]
> doesn't provide native qt style

I figured it was different somehow, that explains it.

> @rnhmjoj Thank you! 

Thank *you*

**Comment by worldofpeace** [comment]
:-1: I feel like at least @jtojnar @ttuegel and myself should have approved on this before it was merged. I even expressed that to @gnidorah. With release management I've had almost no time to look at these things, which I'm sorry about. I probably should have at least mentioned that here also for everyone to see, instead of the mentions I made on IRC.

**Comment by worldofpeace** [comment]
> IMHO this has waited long enough, I hoped it would be in 20.03 already. Let's merge it.

As mentioned, it may look like it has waited and received improvements off feedback, but there really hasn't been any.

**Comment by rnhmjoj** [comment]
> 👎 I feel like at least @jtojnar @ttuegel and myself should have approved on this before it was merged. I even expressed that to @gnidorah.

I'm sorry. I saw the changes you requested were completed and @gnidorah had to fix conflicts several times so I thought it would be better to merge it now.

**Comment by worldofpeace** [comment]
> > -1 I feel like at least @jtojnar @ttuegel and myself should have approved on this before it was merged. I even expressed that to @gnidorah.
> 
> I'm sorry. I saw the changes you requested were completed and @gnidorah had to fix conflicts several times so I thought it would be better to merge it now.

Indeed, it's sad for something like this to go unnoticed.

**Comment by worldofpeace** [comment]
#80854. Could you open another PR for this?
And I'd request review of @jtojnar myself @ttuegel on the qt module.
I'd for sure like to see @Infinisil :heavy_check_mark: on it since it's a new module.

**Comment by worldofpeace** [comment]
It would also help, though I think you'd have to rewrite your history :frowning:, if the gtk module and the qt module weren't in the same PR. It just makes it easier to review.

**Comment by pinpox** [comment]
Is there any way to override theme colors declaratively? I'd like to replace some colors with custom ones I have defined to be used throughout my configuration in variables

**Comment by gnidorah** [comment]
@pixpox You could try something like this. Note that Qt theming is broken due #98009
```nix
  environment.etc."xdg/gtk-2.0/gtkrc".text = ''
    gtk-theme-name="Adwaita-dark"
  '';
  environment.etc."xdg/gtk-3.0/settings.ini".text = ''
    [Settings]
    gtk-theme-name=Adwaita-dark
  '';
  # https://github.com/NixOS/nixpkgs/issues/87667
  environment.variables.QT_QPA_PLATFORMTHEME = "gtk3";
  environment.variables.QT_STYLE_OVERRIDE = "adwaita-dark";
  services.xserver.windowManager.i3.extraPackages =
    options.services.xserver.windowManager.i3.extraPackages.default ++ [ pkgs.adwaita-qt ];
```

**Comment by pinpox** [comment]
Hm, that only sets the the gtk theme, if I'm not mistaken. I was referring to changing the individual colors. For now I ended up doing it like this: https://github.com/pinpox/nixos-home/blob/1cefe28c72930a0aed41c20d254ad4d193a3fa37/gtk.nix#L11 
Maybe someone knows a simpler solution?

**Comment by rycee** [comment]
@pinpox That looks quite nice. Under which license is your configuration? I'd like to include a setup like yours in my the [`gtk.nix`](https://gitlab.com/rycee/nur-expressions/-/blob/master/hm-modules/theme-base16/gtk.nix) file but need it to be compatible with MIT.

**Comment by pinpox** [comment]
@rycee Sure thing, go ahead. I have added MIT to the files, let me know if you have any questions

**Comment by rycee** [comment]
@pinpox Thanks, I've [added it now](https://gitlab.com/rycee/nur-expressions/-/commit/b95d73938318f1a6c81425d85b55441c00a93df5).

**Pull request merged**